db/migrate/20210830183627_remove_classification_id_null_constraint_from_extracts.rb

```diff
@@ -0,0 +1,5 @@
+class RemoveClassificationIdNullConstraintFromExtracts < ActiveRecord::Migration[5.2]
+  def change
+    change_column_null :extracts, :classification_id, true
```

Comment:
> So if we upload an extract for multiple subjects with the same extractor_key, we cannot insert to db, because of this unique constraint.

How will this still work with `classification_id: nil, extractor_key: 'already_set'` won't that still fall foul of the uniq index even with the not null constraint removed?

Reply:
> How will this still work with `classification_id: nil, extractor_key: 'already_set'` won't that still fall foul of the uniq index even with the not null constraint removed?

I'm trying to find a good article that backs up what I've been seeing, but I can't seem to find a good one. What I've been seeing is that the `unique constraint` is skipped if `classification_id` is `nil`, which allows us to insert/upsert extracts. I'm thinking null values aren't compared because they're viewed as having no value, but thats a guess. 

> OOI - was this inserting a classification to the panoptes DB or the caesar DB? In the caesar data model, they don't have to be linked or being consistent with panoptes DB (originally they did as caesar backfilled from Panoptes API but now we push data to caesar for processing).

I think initially there were talks of inserting classification to panoptes DB and which (from what I understood) triggers an event in caesar to go through extractors/then reducers to get reductions (and i think delete classification from caesar db).  I guess we could create the classification in caesar db, but I wonder what would happen if someone reruns extractors. (I understood it as when we rerun extractors, if there's a classification id on extract, we pull the classification  from panoptes and run extractors ?)

> Perhaps for each import we create a special classification (with negative integers PK starting at 0 or -1) record that helps avoid the uniq index issue and avoids removing the not null constraint as well.

In ^ that case, wouldn't we need to update the unique constraint rule with a check that says, "if its this special classification id, skip unique constraint"? (Is that bulkier/heavier than removing not null constraint?)

Reply:
> I'm trying to find a good article that backs up what I've been seeing, but I can't seem to find a good one. What I've been seeing is that the unique constraint is skipped if classification_id is nil, which allows us to insert/upsert extracts. I'm thinking null values aren't compared because they're viewed as having no value, but thats a guess.

Should be easy to test by adding two records with the same index values on your test db and seeing what happens.

Reply:
> Should be easy to test by adding two records with the same index values on your test db and seeing what happens.
Right. Sorry for confusion there. So what I'm seeing is that we can have multiple `classification_id: nil, extractor_key: 'already_set'` extracts, inserting won't fail.  But multiple `classification_id: 29 , extractor_key: 'already_set'` inserts aren't allowed by unique constraint rule. 
I can't give a good explanation why this (i.e. unique constraint rule skips null) happens, I just know that from testing that it happens.

Reply:
>  I guess we could create the classification in caesar db, but I wonder what would happen if someone reruns extractors. (I understood it as when we rerun extractors, if there's a classification id on extract, we pull the classification from panoptes and run extractors ?)

Yep - it attempts to backfill from the main API as outlined here, https://github.com/zooniverse/caesar/blob/c57085b5b43b126635c00cb3f24a7f39556a7cae/app/models/workflow.rb#L152 but it doesn't have to be able to re-run extractors on if we don't want it to (though modifying this behavior is adding unwanted complexity)

OOI - what do you think will happen if you have the not null constraint and someone still presses `re-run extracts`. It will still attempt to backfill the data as outlined above and perhaps cause issues anyway.... 🤔 even with the null constraint removed (it doesn't reflect on the classification ids to backfill, rather it fetches all known classifications for the workflow / subject as outlined https://github.com/zooniverse/panoptes-client.rb/blob/fbe207de73afb1392ff1290b8af6c66ad08ef660/lib/panoptes/client/classifications.rb#L6-L12 as it's called from https://github.com/zooniverse/caesar/blob/c57085b5b43b126635c00cb3f24a7f39556a7cae/app/workers/fetch_classifications_worker.rb#L32

Reply:
> > Should be easy to test by adding two records with the same index values on your test db and seeing what happens.
> 
> Right. Sorry for confusion there. So what I'm seeing is that we can have multiple `classification_id: nil, extractor_key: 'already_set'` extracts, inserting won't fail. But multiple `classification_id: 29 , extractor_key: 'already_set'` inserts aren't allowed by unique constraint rule. I can't give a good explanation why this (i.e. unique constraint rule skips null) happens, I just know that from testing that it happens.

Looks like PG treats NULL values as non matching for uniqueness - https://www.postgresql.org/docs/11/indexes-unique.html
> Null values are not considered equal. A multicolumn unique index will only reject cases where all indexed columns are equal in multiple rows.

TIL 👍🏼

Reply:
> My main concern with removing the not null constraint is allowing our internal data model to break and leaving us with a possibly inconsistent data set. My concerns may very well be ill founded but with the non-null constraint in place, any service pushing data into our system has to conform to match the data model we expect. Removing this constraint allows folks to push data into the system that may break some other part of the system relies upon.

> If need be we can certainly remove this constraint but it comes with a data consistency cost.

I definitely understand the concern. This was something I brought up to Cliff and Zach prior because we definitely had reservations on doing this as well. But chose this way as almost a last resort in order to keep things as simple as we can make it for now. (Knowing that this will have to be looked at more in the future and treating this more as an iteration 1 learning as we go process) 

I'm definitely all ears if there's a better solution, for all the above reasons you mentioned. :) 

> Please test the Extract table unique index with data from multiple Extracts with nil classifications and the same extractor key. If that index fails to insert multiple records then the above point is moot and we will need to consider alternatives.

Sure! I can test that again. Linked is my test csv https://panoptes-uploads.zooniverse.org/project_attached_image/c7208dac-5296-447b-979f-0f4d691a5031.csv. In this case extracts created have the same `extractor_key`, `classification_id` is `nil` and all extracts are getting inserted without failure.